### PR TITLE
Fix/int 859 level

### PR DESCRIPTION
**Pull request by roberto-butti**
## Pull request type

Jira Link: INT-859

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Other (please describe):

## What is the new behavior?

Resolving nested relations.

This PR introduces the feature to resolve nested relations, tracking the level for enrichContent() method

For testing:
```
composer run test
```
Especially take a look at tests:
- Nested relations ( less than 50): "Integration: get one story with few resolved relations" 
- Nested relations: "Integration: get one story with Resolved relations 1"
- Nested relations (more than 50):"Integration: get one story with Resolved relations 2"
- Nested translated relations :"Integration: get list of stories -translations- story with Resolved relations 2"
- Nested relatins for multiple stories: "Integration: get list of stories story with Resolved relations 2"
- Auto referenced stories (avoid infinite loop): "Integration: test stop resolving loop"
- 
in the file `tests/Storyblok/Integration/ClientTest.php`

### CODE REVIEW

**Review by joaokamun — APPROVED** [review]

File: tests/Storyblok/Integration/ClientTest.php [hunk 1]
```diff
@@ -154,6 +151,52 @@
...
+    $key = 'stories/' . $slug;
+    $client->editMode();
+    $options = $client->getApiParameters();
+    $client->resolveRelations(
+        'Product.ProductVariants'
+    );
+    $responses = $client->getStoryBySlug($slug);
+    $body = $responses->getBody();
+    expect($body)->toHaveKeys(['rels', 'story', 'cv', 'links']);
+    // print_r($body);
```
Comment (ademarCardoso) [inline tests/Storyblok/Integration/ClientTest.php:191]: Please, remove this comment

**Comment by ademarCardoso** [comment]
Hello my friend, i have a test case of a client (it was not the one who reported the problem) that has nested stories, and the php client still does not generate the correct json, in the tests with the new script the generated json has 492 lines, in the js- client generates around 1800 lines, as all relationships are resolved, below is the script I used in PHP and JS-client.

```php
<?php

require './vendor/autoload.php';

$client = new \Storyblok\Client('HNDqMoMAwQD5l06WBEGTUwtt');

$slug = '/da/services/service-page-2';

$client->editMode();

$options = $client->getApiParameters();
$client->resolveRelations(
    'Global.globalref,ServicSidebar.pages'
);
$responses = $client->getStoryBySlug($slug);
$body = $responses->getBody();


$path = './relations-by-php.json';
$jsonString = json_encode($responses, JSON_PRETTY_PRINT);
$fp = fopen($path, 'w');
fwrite($fp, $jsonString);
fclose($fp);
```

```js
import StoryblokClient from 'storyblok-js-client'
import fs from 'fs'

const relationsResolvers = ['Global.globalref', 'ServicSidebar.pages']
const slug2 = '/services/service-page-2'
const locale = '/da'
const jsClientVersion = 'V5-json'

try {
  let sbApi = new StoryblokClient({
    accessToken: 'HNDqMoMAwQD5l06WBEGTUwtt',
    version: "draft",
  })

  await sbApi.get(`cdn/stories${locale}${slug2}`, { version: "draft", resolve_relations: relationsResolvers })
    .then((response) => {
      console.log('Works!')

      fs.writeFile(`${jsClientVersion}.json`, JSON.stringify(response.data, null, 2), function(err) {
          if (err) {
              console.log(err);
          }
      });
    })
    .catch((error) => {
      console.log('Error =>', error)
    })
} catch (error) {
  console.log('Error try =>', error)
}

```

**Comment by ebrunaway** [comment]
we tried the new SDK but we found an issue with Single Option fields: in some case we get an object returned (which we think is correct) in other cases we get an array of one element containing the object. Maybe it has to do with multiple levels of nesting. Look at the attachment in the field `parent`: on the right we get an array of 1 element with the content inside, in the other case (the left) we get the content directly resolved. The right side is using the new SDK, the left one is using the old SDK.

Screenshot 2023-04-11 at 13 01 50

**Comment by ebrunaway** [comment]
@ademarCardoso @joaokamun thanks for the update, we tested the resolution of nested levels and now it solves properly everything! The bug we had in the previous screenshot is gone, unfortunately we have a similar issue on another `singleOption`. See the screen attached:
Screenshot 2023-04-12 at 10 18 27
on the left we're using the new SDK implementation, on the right we have the old one. On the left the `singleOption` - which we expect to be an object - is returned as an array of 1 element. In the old SDK it's solved correctly as an object. We're not sure why this is happening because for other singleOptions it appears to be fixed, but probably there must be some edge cases where it stills break.

**Comment by ebrunaway** [comment]
Maybe this can help, we've found an occurrence of this bug within `singleOption` inside a `multiOption` array. The `multiOption` is an array of 5 elements who's content it's a `singleOption` story. See the screen attached:
Screenshot 2023-04-12 at 16 45 13
The items from the 0 to the 3rd element are resolved correctly as an object, the 4th (which is also the last item in the array) is solved as array, but we were expecting to receive an obj.

**Comment by joaokamun** [comment]
@ebrunaway please can you check if my last commit is resolving correctly singleOptions on a multiOption array?

**Comment by ebrunaway** [comment]
Hi @joaokamun thanks for your help with this! It looks like `singleOption` are now solved but the fix introduced two problems (hopefully easy to fix!):

1. now the elements inside `multiOptions` are resolved except the last element of the list. Look at this screenshot:
Screenshot 2023-04-13 at 09 31 22

We have a component called `popupDataVariation` that is a `multiOptions` containing a `singleOption` called `boxNumbers` which have inside 6 `box`. The elements in `popupDataVariation` from 0 to 3 are resolved properly whilst the 4th is returning us `uuids`. This is happening even with other stories so it's not an isolated behaviour and it's always presenting in the last element of the array.

2. the `multiOptions` of 1 element now are returning an object instead of an array of 1 element. The expected behavior should be that a `multiOptions` must always return an array (independently of how many items it contains) and the `singleOption` must always return an object (this is working now). See the screen attached:
MicrosoftTeams-image (6)

here we have `cardAndPopup` which is a `multiOption` with one element inside, but as you can see from the screenshot it's returning us an object instead of an array.

**Comment by ebrunaway** [comment]
looks like the last commit fixed both issues 👍

**Comment by joaokamun** [comment]
Thanks @ebrunaway , I still going deeper in other scenarios to check if everything is working as expected before merging this PR.

**Pull request merged**